From ea47d6c8b13d7a755318674269a7de9dc0d9f8cb Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Fri, 4 Nov 2016 11:13:02 +0000
Subject: [PATCH 23/42] Win32 (MSVC): Cannot use variadic token pasting GNU ext

I have read [1] that MSVC gets rid of any trailing ',' but I have not
seen evidence (*) of this so I dropped comma before the __VA_ARGS__.
This may result in the __FILE__ argument being combine with the first
variadic argument.

This change is untested (except that it compiles in the case where it was
broken) because I cannot make a debug build without debug Boost libraries
(and probably debug Qt libraries too).

There are cross-platform approaches [1] that could be used if needs be.

[1] http://stackoverflow.com/questions/5588855/standard-alternative-to-gccs-va-args-trick

* This commit assumes that what MSVC does is to insert a ',' when needed.
---
 src/cpp/tests/cpp/tests/TestThat.hpp | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/cpp/tests/cpp/tests/TestThat.hpp b/src/cpp/tests/cpp/tests/TestThat.hpp
index c0952b8720..7633989a73 100644
--- a/src/cpp/tests/cpp/tests/TestThat.hpp
+++ b/src/cpp/tests/cpp/tests/TestThat.hpp
@@ -22,7 +22,11 @@
 
 # ifndef RSTUDIO_NO_TESTTHAT_ALIASES
 
-#  define context(__X__, ...) TEST_CASE(__X__, __FILE__)
+#ifdef _MSC_VER
+#  define context(__X__, ...) TEST_CASE(__X__, __FILE__  __VA_ARGS__)
+#else
+#  define context(__X__, ...) TEST_CASE(__X__, __FILE__, ##__VA_ARGS__)
+#endif
 #  define test_that SECTION
 #  define expect_true(x) CHECK((x))
 #  define expect_false(x) CHECK_FALSE((x))
-- 
2.14.3 (Apple Git-98)

